### PR TITLE
layers: Add VK_KHR_dynamic_rendering VUs and Tests

### PR DESCRIPTION
**Pull request by AaronHaganAMD**
VU Statements:
VUID-VkGraphicsPipelineCreateInfo-renderPass-06053
VUID-VkGraphicsPipelineCreateInfo-renderPass-06054
VUID-VkRenderingAttachmentInfoKHR-imageView-06145
VUID-VkRenderingAttachmentInfoKHR-imageView-06146
VUID-VkRenderingInfoKHR-imageView-06107
VUID-VkRenderingInfoKHR-imageView-06108
VUID-VkRenderingInfoKHR-imageView-06109
VUID-vkCmdDrawMeshTasksIndirectNV-*
VUID-vkCmdDrawMeshTasksIndirectCountNV-*
VUID-vkCmdDrawIndirectByteCountEXT-*

Tests:
VUID-VkGraphicsPipelineCreateInfo-renderPass-06053
VUID-VkGraphicsPipelineCreateInfo-renderPass-06054
VUID-VkRenderingAttachmentInfoKHR-imageView-06129
VUID-VkRenderingAttachmentInfoKHR-imageView-06132
VUID-VkRenderingAttachmentInfoKHR-imageView-06135
VUID-VkRenderingAttachmentInfoKHR-imageView-06136
VUID-VkRenderingAttachmentInfoKHR-imageView-06145
VUID-VkRenderingAttachmentInfoKHR-imageView-06146
VUID-VkRenderingInfoKHR-imageView-06107
VUID-VkRenderingInfoKHR-imageView-06108
VUID-VkRenderingInfoKHR-imageView-06116
VUID-VkRenderingInfoKHR-viewMask-06069
VUID-VkRenderingInfoKHR-multiview-06127

### CODE REVIEW

**Comment by ncesario-lunarg** [comment]
I haven't had a chance to review this in-depth, but looks like:
- More copyright updates necessary for the new year 🎉 
- `VkLayerTest.DynamicRenderingAttachmentInfo` is crashing on nvidia HW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -2600,6 +2600,28 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
...
     const auto rendering_struct = LvlFindInChain<VkPipelineRenderingCreateInfoKHR>(create_info.pNext);
     if (rendering_struct) {
+        if ((pPipeline->active_shaders & VK_SHADER_STAGE_FRAGMENT_BIT) &&
+            ((rendering_struct->depthAttachmentFormat != VK_FORMAT_UNDEFINED) ||
+             (rendering_struct->stencilAttachmentFormat != VK_FORMAT_UNDEFINED)) &&
+            ((create_info.pDepthStencilState == nullptr) ||
+             (create_info.pDepthStencilState->sType != VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO))) {
+            skip |= LogError(
+                device, "VUID-VkGraphicsPipelineCreateInfo-renderPass-06053",
+                "Pipeline %" PRIu32
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:2610]: This should start with the name of the API that triggered the error, so I think something like `vkCreateGraphicsPipelines: ...` in this case.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -6841,9 +6864,65 @@ bool CoreChecks::PreCallValidateCmdBeginRenderingKHR(VkCommandBuffer commandBuff
...
+                    if (!(image_view_state->create_info.flags & VK_IMAGE_CREATE_SUBSAMPLED_BIT_EXT)) {
+                        skip |= LogError(
+                            commandBuffer, "VUID-VkRenderingInfoKHR-imageView-06107",
+                            "vkCmdBeginRenderingKHR(): color image must be created with VK_IMAGE_CREATE_SUBSAMPLED_BIT_EXT");
+                    }
+                }
+            }
+
+            if (pRenderingInfo->pDepthAttachment) {
+                auto depth_view_state = Get<IMAGE_VIEW_STATE>(pRenderingInfo->pDepthAttachment->imageView);
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:6880]: Looks like `imageView` is optional for `VkRenderingAttachmentInfoKHR`? Should we check to make sure it's not `VK_NULL_HANDLE` (or similarly that `depth_view_state != nullptr`)?

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -6841,9 +6864,65 @@ bool CoreChecks::PreCallValidateCmdBeginRenderingKHR(VkCommandBuffer commandBuff
...
+
+        if (fragment_density_map_attachment_info->imageView != VK_NULL_HANDLE) {
+            auto fragment_density_map_view_state = Get<IMAGE_VIEW_STATE>(fragment_density_map_attachment_info->imageView);
+
+            if ((pRenderingInfo->viewMask == 0) &&
+                (fragment_density_map_view_state->create_info.subresourceRange.layerCount != 1)) {
+                skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-imageView-06109",
+                                 "vkCmdBeginRenderingKHR(): imageView of VkRenderingFragmentDensityMapAttachmentInfoEXT must "
+                                 "have a laycount ("
+                                 "%" PRIu32 ") equal to 1 when viewMask 0",
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:6913]: "viewMask `is`/`=` 0`?

**Review by ncesario-lunarg — CHANGES_REQUESTED** [review]
Some minor nits, but overall LGTM!

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -2600,6 +2600,28 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
 
     const auto rendering_struct = LvlFindInChain<VkPipelineRenderingCreateInfoKHR>(create_info.pNext);
     if (rendering_struct) {
```
[the remark this answers is not in the document]
Comment (AaronHaganAMD) [reply layers/core_validation.cpp:2602]: if VkPipelineRenderingCreateInfoKHR is used, that would imply renderPass is VK_NULL_HANDLE.

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -2600,6 +2600,28 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
...
+        if ((pPipeline->active_shaders & VK_SHADER_STAGE_FRAGMENT_BIT) &&
+            ((rendering_struct->depthAttachmentFormat != VK_FORMAT_UNDEFINED) ||
+             (rendering_struct->stencilAttachmentFormat != VK_FORMAT_UNDEFINED)) &&
+            ((create_info.pDepthStencilState == nullptr) ||
+             (create_info.pDepthStencilState->sType != VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO))) {
```
[the remark this answers is not in the document]
Comment (AaronHaganAMD) [reply layers/core_validation.cpp:2607]: ` ((create_info.pDepthStencilState == nullptr) ||
             (create_info.pDepthStencilState->sType != VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO)))`

Was supposed to be the check for `and an invalid pDepthStencilState structure` of the VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 validation statement.  Do you suggest just checking against null to see if its a valid pDepthStencilState structure ?

[the remark this answers is not in the document]
Comment (ncesario-lunarg) [reply layers/core_validation.cpp:2607]: Correct, that _should_ be sufficient, as I'm pretty sure the `sType` check should have already occurred during parameter validation.

File: tests/vklayertests_dynamic_rendering.cpp [hunk 6]
```diff
@@ -277,4 +281,185 @@ TEST_F(VkLayerTest, DynamicRenderingGraphicsPipelineCreateInfo) {
...
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingInfoKHR-imageView-06107");
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingInfoKHR-imageView-06108");
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingInfoKHR-imageView-06116");
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingAttachmentInfoKHR-imageView-06129");
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingAttachmentInfoKHR-imageView-06145");
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkRenderingAttachmentInfoKHR-imageView-06146");
+    m_commandBuffer->BeginRendering(begin_rendering_info);
+    m_errorMonitor->VerifyFound();
+    m_commandBuffer->EndRendering();
+    m_commandBuffer->end();
```
Comment (ncesario-lunarg) [inline tests/vklayertests_dynamic_rendering.cpp:405]: @AaronHaganAMD I think these calls (and similarly for the following test) need to be removed. It is crashing in the nvidia driver, likely due to the fact that vkCmdBeginRenderingKHR never gets dispatched to the driver due to the validation errors.

**Comment by ncesario-lunarg** [comment]
Crashes fixed, now all tests are passing except 1:
```
[ RUN      ] VkLayerTest.InvalidPipelineRenderingParameters
Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_R8G8B8A8_UNORM) or stencil format (VK_FORMAT_UNDEFINED) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: FailedReceived unexpected error(s).
Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_UNDEFINED) or stencil format (VK_FORMAT_R8G8B8A8_UNORM) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: Failed
Received unexpected error(s).
Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_D32_SFLOAT) or stencil format (VK_FORMAT_D32_SFLOAT_S8_UINT) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: Failed
```

Looks like this is a different dynamic rendering test added by @stu-s?

**Comment by AaronHaganAMD** [comment]
> Crashes fixed, now all tests are passing except 1:
> 
> ```
> [ RUN      ] VkLayerTest.InvalidPipelineRenderingParameters
> Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_R8G8B8A8_UNORM) or stencil format (VK_FORMAT_UNDEFINED) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
> C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: FailedReceived unexpected error(s).
> Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_UNDEFINED) or stencil format (VK_FORMAT_R8G8B8A8_UNORM) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
> C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: Failed
> Received unexpected error(s).
> Unexpected: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-renderPass-06053 ] Object 0: handle = 0x2061e79b480, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x247dedb2 | vkCreateGraphicsPipelines(): Pipeline 0 has fragment state and a depth format (VK_FORMAT_D32_SFLOAT) or stencil format (VK_FORMAT_D32_SFLOAT_S8_UINT) and an invalid pDepthStencilState structure The Vulkan spec states: If renderPass is VK_NULL_HANDLE, the pipeline is being created with fragment shader state, and either of VkPipelineRenderingCreateInfoKHR::depthAttachmentFormat or VkPipelineRenderingCreateInfoKHR::stencilAttachmentFormat are not VK_FORMAT_UNDEFINED, pDepthStencilState must be a valid pointer to a valid VkPipelineDepthStencilStateCreateInfo structure (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-renderPass-06053)
> C:\lunarg\external-prs\Vulkan-ValidationLayers\tests\vkrenderframework.cpp(206): error: Failed
> ```
> 
> Looks like this is a different dynamic rendering test added by @stu-s?

VkLayerTest.InvalidPipelineRenderingParameters should be passing now...

**Comment by stu-s** [comment]
@AaronHaganAMD / @ncesario-lunarg is this ok to be merged now?

**Comment by AaronHaganAMD** [comment]
> @AaronHaganAMD / @ncesario-lunarg is this ok to be merged now?

I believe everything has been addressed.

**Review by ncesario-lunarg — APPROVED** [review]

**Pull request merged**